rhp/v4/transport.go

```diff
@@ -18,16 +18,16 @@ func withDecoder(r io.Reader, maxLen int, fn func(*types.Decoder)) error {
 	return d.Err()
 }
 
-// ReadID reads an RPC ID from the stream.
-func ReadID(r io.Reader) (id types.Specifier, err error) {
-	err = withDecoder(r, 16, id.DecodeFrom)
+// ReadID reads an RPC header from the stream.
+func ReadID(r io.Reader) (id RPCSpecifier, err error) {
+	err = withDecoder(r, 16, (*types.Specifier)(&id).DecodeFrom)
 	return
```

Comment:
Shouldn't we also check that the combination of RPC and version is a valid one here? 
It's usually a good idea to only construct and return valid instances of a type when parsing data.

Reply:
Tbh it doesn't matter too much. The whole `transport.go` file should probably live in `coreutils` anyway since it is a utility file that is neither used nor tested in `core`. But I'd argue that if we don't perform validation, it would be better to return `types.Specifier` here. If I call `ParseHTTPMethod` of some `http-core` library I'd also expect an error if it is a non-std method. 

Makes me wish this was Rust. Because then `RPCSpecifier` would probably be an enum wrapping a `Specifier` and to return it we'd be forced to validate it here. Which would then nicely force all devs to exhaustively switch over all possible variations whenever they update the dependency.

Reply:
You love to watch the world burn don't you?